Enable Serialization/Deserialization round-trip for Domain objects

jeffmcadams opened the pull request:
The custom validator on the 'created' and 'expires' fields on the Domain object allowed for either a datetime object directly, or Namecheap's 'MM/DD/YYYY' string format. But if you serialize the object (ie, model_dump_json()) and then try to deserialize it using the Domain model, it would fail because those fields would have strings of the format 'YYYY-MM-DDThh-mm-ss' which wouldn't validate.

This adds a quick if/else to see if there is a 'T' in the datetimestamp string, if so it uses datetime.fromisoformat(v), rather than the datetime.strptime(v, "%m/%d/%Y').

This change allows a Domain object to be serialized to JSON, and reserialized back into a Domain object.

jeffmca committed:
Enable Serialization/Deserialization round-trip for Domain objects

The custom validator on the 'created' and 'expires' field allowed for
either a datetime object directly, or Namecheap's 'MM/DD/YYYY' string
format. But if you serialize the object (ie, model_dump_json()) and then
try to deserialize it using the Domain model, it would fail because
those fields would have strings of the format 'YYYY-MM-DDThh-mm-ss'
which wouldn't validate.

This adds a quick if/else to see if there is a 'T' in the datetimestamp
string, if so it uses datetime.fromisoformat(v), rather than the
datetime.strptime(v, "%m/%d/%Y').

This change allows a Domain object to be serialized to JSON, and
reserialized back into a Domain object.

adriangalilea commented:
Hey @jeffmcadams! Thanks for tackling the Pydantic round-trip serialization issue.

**Issue with current implementation:**

The logic has a bug that would break existing Namecheap API parsing:

```python
if (v.find('T') == 01):  # Checks if 'T' is at index 1, not if it exists!
    return datetime.strptime(v, "%m/%d/%Y")
else:
    return datetime.fromisoformat(v)
```

`v.find('T') == 01` checks if `'T'` is at **index 1**, not if it exists (returns -1 when not found). Tested: Namecheap format `'01/15/2024'` fails with "Invalid isoformat string". This would break all existing Domain parsing from the Namecheap API.

**Suggested fix:**

```python
@field_validator("created", "expires", mode="before")
@classmethod
def parse_datetime(cls, v: Any) -> datetime:
    """Parse datetime from Namecheap API or Pydantic serialization."""
    if isinstance(v, datetime):
        return v
    if isinstance(v, str):
        # Namecheap API format (MM/DD/YYYY)
        if '/' in v:
            return datetime.strptime(v, "%m/%d/%Y")
        # Pydantic serialization format (ISO 8601)
        return datetime.fromisoformat(v)
    raise ValueError(f"Cannot parse datetime from {v}")
```

This handles both formats correctly:
- Namecheap API: `'01/15/2024'` parsed via `strptime`
- Pydantic JSON: `'2024-01-15T14:30:00'` parsed via `fromisoformat`
- Datetime objects: pass through unchanged

I can commit this directly to your PR if you agree, but I'll wait for your thoughts first.

jeffmcadams commented:
Oops, I made a typo as a copied across my changes from one repo to another (I know, not great workflow on my part...but it's what I have to deal with for the moment).

It was supposed to be "-1" instead of "01".

I kinda think looking for the "T" and using the iso format and then "falling back" to the Namecheap format would be the more robust approach, but it's also backwards from what's probably the primary use case for the code, so ultimately I don't have much of an opinion on which alternative to use. They really should behave the same with the set of potential inputs I'm thinking of.

adriangalilea committed:
fix: correct datetime parser to handle both Namecheap and Pydantic formats

The previous logic had inverted conditions that would break Namecheap API
parsing. Now properly detects format using '/' character:
- Namecheap API: MM/DD/YYYY (has '/')
- Pydantic JSON: ISO 8601 (no '/')

This enables Domain objects to be serialized to JSON and deserialized back.

adriangalilea commented:
Pushed the fix to your branch! Used the `'/' in v` check to detect Namecheap format, then fallback to ISO.

Tested:
- ✓ Namecheap API format (`'01/15/2024'`)
- ✓ Pydantic JSON format (`'2024-01-15T14:30:00'`)
- ✓ Full round-trip serialization/deserialization

Thanks for the contribution!

jeffmcadams commented:
Awesome...just double-checked on my setup and it does work, as expected.

I haven't dug too deeply into your library yet, if I find any other things like this, I'll be sure to send a PR your way

Thanks!

adriangalilea commented:
> Awesome...just double-checked on my setup and it does work, as expected.
> 
> I haven't dug too deeply into your library yet, if I find any other things like this, I'll be sure to send a PR your way
> 
> Thanks!

I didn’t have time to work on it much beyond the initial sprint so any feedback or proposal is welcome.

Happy that is working for you!
